### PR TITLE
homework_2

### PR DESCRIPTION
**Pull request by lysiakin**

### CODE REVIEW

File: src/Controller/Api/ApiController.php [hunk 1]
```diff
@@ -2,107 +2,81 @@
...
     {
-        if (!$request->headers->get('Content-type: application/json')) {
-            throw new UnsupportedMediaTypeHttpException();
-        }
+        $clientInfo = [
+            'ip' => $request->getClientIps()[0],
+            'language' => $request->getPreferredLanguage(),
+            'browser' => $request->headers->get('User-Agent')
+        ];
+        $encoded = json_encode($clientInfo);
```
Comment (anyt) [inline src/Controller/Api/ApiController.php:45]: there is no need to encode an array, if it's rendered with the JsonResponse

File: src/Controller/Api/ApiController.php [hunk 2]
```diff
@@ -2,107 +2,81 @@
...
-        });
+        $urls = $this->generateUrls($data, $urlGenerator);
+        $neededProducts = array_filter($urls, fn ($item) => str_contains($item, $product));
 
         if (empty($neededProducts)) {
             throw new NotFoundHttpException();
         }
 
         $neededProduct = $data[key($neededProducts)];
+        $translatedKeys = $this->translate($this->getTranslateIds(), $translator);
```
Comment (anyt) [inline src/Controller/Api/ApiController.php:77]: Having two separate methods that must be running in a chain is not the best design.
It's much better to inline the code from getTranslateIds here, as it's easier to read the code without scrolling back and forth. Too small methods to follow SRP is a bad practice.

File: src/Controller/Api/ApiController.php [hunk 3]
```diff
@@ -121,21 +95,23 @@ private function getDataFromFile(): array
...
      * Returns formatted list of resources.
      *
-     * @param Request $request
      * @param array $data
+     * @param UrlGeneratorInterface $urlGenerator
      *
      * @return array
      */
-    private function prepareData(Request $request, array $data): array
+    private function generateUrls(array $data, UrlGeneratorInterface $urlGenerator): array
```
Comment (anyt) [inline src/Controller/Api/ApiController.php:103]: This interface is not clear. Try avoding usage of plain arrays in the code as it's hard to maintein for other developers. An interface must be clear without the need to check the implementation.

File: src/Controller/Api/ApiController.php [hunk 4]
```diff
@@ -155,11 +131,35 @@ private function formatUri(string $product): string
...
+    
+    /**      
+     * Translates keys of the Product object.
+     * 
+     * @param array $translateIdentifiers
+     * @param $translator  
+     * 
+     * @return array
+     */
+    private function translate(array $translateIdentifiers, $translator) : array
```
Comment (anyt) [inline src/Controller/Api/ApiController.php:150]: The method is redundant
- the translator interface is not specified
- it's not clear what should be transferred to translateIdentifiers
- and the presence of this method forces you to iterate over an array two times, at first to get Ids, then to translate them.

**Review by anyt — CHANGES_REQUESTED** [review]

File: src/Controller/Api/ApiController.php (hunk 2 above)
Comment (lysiakin) [reply src/Controller/Api/ApiController.php:77]: I assumed that i recieve translate ids from database in getTranslateIds() method. So thats only the reason.

Comment (lysiakin) [reply src/Controller/Api/ApiController.php:77]: But anyway thanks, i'm appreciate it

File: src/Controller/Api/ApiController.php (hunk 1 above)
Comment (lysiakin) [reply src/Controller/Api/ApiController.php:45]: It's just for logging purposes which are at 47

**Pull request merged**